bluepill/src/BPPacker.m

```diff
@@ -86,5 +98,77 @@ @implementation BPPacker
     return bundles;
 }
 
++ (NSMutableArray<BPXCTestFile *> *)packTestsByTime:(NSArray<BPXCTestFile *> *)xcTestFiles
+                                      configuration:(BPConfiguration *)config
+                                           andError:(NSError **)errPtr {
+    [BPUtils printInfo:INFO withString:@"Packing based on individual test execution times in file path: %@", config.testTimeEstimatesJsonFile];
+    if (xcTestFiles.count == 0) {
+        BP_SET_ERROR(errPtr, @"Found no XCTest files.\n"
+                     "Perhaps you forgot to 'build-for-testing'? (Cmd + Shift + U) in Xcode.");
+        return NULL;
+    }
+
+    // load the config file
+    NSError *error;
+    NSDictionary *testTimes = [self loadConfigFile:config.testTimeEstimatesJsonFile withError:&error];
+    if (!testTimes) {
+        BP_SET_ERROR(errPtr, @"Could not load configuration from '%@'\n%@", optarg, [error localizedDescription]);
+        return NULL;
+    }
+
+    NSMutableDictionary *estimatedTimesByTestFilePath = [[NSMutableDictionary alloc] init];
+    NSArray *testCasesToRun = config.testCasesToRun;
 
+    double totalTime = 0.0;
+    for (BPXCTestFile *xctFile in xcTestFiles) {
+        NSMutableSet *bundleTestsToRun = [[NSMutableSet alloc] initWithArray:[xctFile allTestCases]];
+        if (testCasesToRun) {
+            [bundleTestsToRun intersectSet:[[NSSet alloc] initWithArray:testCasesToRun]];
+        }
+        if (config.testCasesToSkip && [config.testCasesToSkip count] > 0) {
+            [bundleTestsToRun minusSet:[[NSSet alloc] initWithArray:config.testCasesToSkip]];
+        }
+        if (bundleTestsToRun.count > 0) {
+            double testBundleExecutionTime = 0.0;
+            for (NSString *test in bundleTestsToRun) {
+                if ([testTimes objectForKey:test]) {
+                    testBundleExecutionTime += [[testTimes objectForKey:test] doubleValue];
+                } else {
+                    [BPUtils printInfo:DEBUGINFO withString:@"Estimated test execution time not found for %@. Settling for a default.", test];
+                    testBundleExecutionTime += 1.0;
```

Comment:
I'm not sure this is a realistic default... maybe we should keep an average running time of the existing tests and use that as the default.

Reply:
Thought of it. Average running time in that set is a slightly better estimate, instead of a random 1.0 seconds. However, that might need a second parse which should be fine. I will incorporate this.

Reply:
Actually, it doesn't need 2 parses if we maintain a count of "known and unknown tests" it can be extrapolated. However, the average execution time is not a sensible default for unknown tests and it could pollute the data.
Also, I noticed that Bluepill's parsing logic considers all methods and even properties with a *test* prefix are considered as test-cases. That's one more reason the data would get polluted by extrapolating with the average.
So, considering ZERO as a sensible default and adding a TODO to revisit.